### PR TITLE
Enhance simplifier to simplify comparison between a constant and a tree of select nodes

### PR DESCRIPTION
**Pull request by rpshukla**
This adds a simplification to a pattern of trees which will be later generated by an optimization in openj9. The code was originally authored by @andrewcraik and was refined by me.

This code further simplifies a pattern of trees that will be generated by the optimization in this PR: https://github.com/eclipse/openj9/pull/8780

This PR should be merged before the openj9 PR so that the openj9 changes can have the maximum benefit to performance.

There are 2 stages in this new simplification:

1. The first stage simplifies comparison nodes where one child is a constant
   and the other is a tree of select nodes with constant leaves.
   The constant is compared to the leaves of the select node(s) and the values
   of these leaves are replaced by zero or one depending on the result of the
   comparison.

2. select nodes with children that are zeros and ones are replaced with simpler
   boolean comparisons.

These changes were originally authored by Andrew Craik who is listed as a
co-author below.

Co-authored-by: Andrew Craik <ajcraik@ca.ibm.com>

Signed-off-by: Ryan Shukla <ryans@ibm.com>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
@cathyzhyi : would you mind reviewing, or suggest a delegate reviewer as appropriate please?

**Comment by andrewcraik** [comment]
Given I helped write this code I am going to refrain from reviewing so I think a review from @cathyzhyi or @vijaysun-omr would be ideal if possible.

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 1]
```diff
@@ -5999,6 +6232,12 @@ TR::Node *treetopSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier *
          }
       node->setFirst(child);
       }
+   if (!node->getFirstChild()->getOpCode().isNullCheck() && node->getFirstChild()->getOpCodeValue() == TR::PassThrough)
```
Comment (cathyzhyi) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: Should this be `!node->getOpCode().isNullCheck()` ?

File: compiler/optimizer/OMRSimplifierHandlers.cpp [hunk 2]
```diff
@@ -13045,6 +13284,11 @@ TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
 
 TR::Node *ificmpeqSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    {
+   // Perform a simplification for the case where an iselect is conpared to a
```
Comment (cathyzhyi) [inline compiler/optimizer/OMRSimplifierHandlers.cpp]: typo `conpared `

File: compiler/optimizer/OMRSimplifierHandlers.cpp (hunk 1 above)
Comment (rpshukla) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: I think so. @andrewcraik could you confirm?

Comment (andrewcraik) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: yup should be what @cathyzhyi suggested.

Comment (rpshukla) [reply compiler/optimizer/OMRSimplifierHandlers.cpp]: Fixed in 805fc5b

**Comment by cathyzhyi** [comment]
LGTM

**Comment by andrewcraik** [comment]
@genie-omr build all

**Comment by rpshukla** [comment]
I'm looking into the test failures. Looks like something went wrong with types for select nodes: `Child 0 has unexpected type Int32 (expected Int64)`

**Comment by rpshukla** [comment]
In `selectSimplifier` I added another condition to the if statement surrounding the code to fold away the select node:
```
        && node->getDataType() == node->getFirstChild()->getDataType())
```

This ensures that that select nodes are only replaced with their first child (the condition node) if the condition node has the same data type as the original select node. This fixed the previously failing test (SelectTest) when I ran it locally.

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
I approve the change, the appveyor failure is'nt a real issue. Merging.

**Pull request merged**